gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java

```diff
@@ -191,59 +197,162 @@ private String dependencyPathsOfProblematicJars(
+    Set<ResolvedComponentResult> checkedCircularDependency = new HashSet<>();
+
+    for (String targetCoordinates : targetCoordinatesSet) {
+      // Queue of dependnecy nodes. Each node knows its parent.
+      ArrayDeque<ResolvedComponentResultNode> queue = new ArrayDeque<>();
+      ResolvedComponentResultNode firstItem =
+          new ResolvedComponentResultNode(componentResult, null);
+      queue.add(firstItem);
+
+      // Mapping to omit duplicate dependency paths in the output. This is a mapping from Maven
+      // coordinates to dependency nodes that has the dependency of the coordinate in their direct
+      // or transitive dependencies.
+      Set<ResolvedComponentResult> nodesDependOnTarget = new HashSet<>();
+
+      while (!queue.isEmpty()) {
+        ResolvedComponentResultNode node = queue.poll();
+        ResolvedComponentResult item = node.componentResult;
+
+        ModuleVersionIdentifier identifier = item.getModuleVersion();
+        String coordinates =
+            String.format(
+                "%s:%s:%s", identifier.getGroup(), identifier.getName(), identifier.getVersion());
+        if (targetCoordinates.equals(coordinates)) {
+          String dependencyPath = node.pathFromRoot();
+          coordinatesToDependencyPaths.put(targetCoordinates, dependencyPath);
+
+          for (ResolvedComponentResultNode iter = node.parent; iter != null; iter = iter.parent) {
+            nodesDependOnTarget.add(iter.componentResult);
+          }
+        }
+
+        if (nodesDependOnTarget.contains(item)) {
+          // Do not show duplicate dependency paths. If we know that this node contains dependency
+          // having targetCoordinates in its direct/transitive dependencies, there is no need to
+          // print the dependency paths again.
+          String dependencyPath = node.pathFromRoot() + " (omitted for duplicate)";
+          coordinatesToDependencyPaths.put(targetCoordinates, dependencyPath);
+
+          for (ResolvedComponentResultNode iter = node; iter != null; iter = iter.parent) {
+            nodesDependOnTarget.add(iter.componentResult);
+          }
+          continue;
+        }
+
+        for (DependencyResult dependencyResult : item.getDependencies()) {
+          if (dependencyResult instanceof ResolvedDependencyResult) {
+            ResolvedDependencyResult resolvedDependencyResult =
+                (ResolvedDependencyResult) dependencyResult;
+            ResolvedComponentResult child = resolvedDependencyResult.getSelected();
+
+            if (node.hasParent(child)) {
+              // Circular dependency check
+              if (checkedCircularDependency.add(child)) {
+                getLogger()
+                    .error(
+                        "Circular dependency for: "
+                            + resolvedDependencyResult
+                            + "\n The stack is: "
+                            + node.pathFromRoot());
+              }
```

Comment:
I'd try to extract this into a `causesCircularDependency()` or similarly named helper. Just generally, the overall BFS structure will be easier to read if the actually useful pieces of functionality were in helper methods.

Reply:
Extracted the bigger for-loop into `getDependencies`.

Reply:
Renamed getDependencies to findDependencies.